open_spiel/games/checkers.cc

```diff
@@ -0,0 +1,546 @@
+
+  // Add an extra space to the bottom row
+  // if the row labels take up two spaces.
+  if (rows_ >= 10) {
+    absl::StrAppend(&result, " ");
+  }
+  absl::StrAppend(&result, " ");
+
+  for (int c = 0; c < columns_; c++) {
+    absl::StrAppend(&result, ColumnLabel(c));
+  }
+  absl::StrAppend(&result, "\n");
+
+  return result;
+}
+
+int CheckersState::ObservationPlane(CellState state, Player player) const {
+  int state_value;
+  switch (state) {
+    case CellState::kWhite:
+      state_value = 0;
+      break;
+    case CellState::kWhiteCrowned:
+      state_value = 1;
+      break;
+    case CellState::kBlackCrowned:
+      state_value = 2;
+      break;
+    case CellState::kBlack:
+      state_value = 3;
+      break;
+    case CellState::kEmpty:
+    default:
+      return 4;
+  }
+  if (player == Player{0}) {
+    return state_value;
+  } else {
+    return 3 - state_value;
+  }
+}
+
+bool CheckersState::IsTerminal() const { 
+  return LegalActions().empty();
+}
+
+std::vector<double> CheckersState::Returns() const {
+  if (outcome_ == kInvalidPlayer || moves_without_capture_ >= kMaxMovesWithoutCapture) {
+    return {0., 0.};
+  } else if (outcome_ == Player{0}) {
+    return {1.0, -1.0};
+  } else if (outcome_ == Player{1}){
+    return {-1.0, 1.0};
+  }
+  return {0., 0.};
+}
+
+std::string CheckersState::InformationStateString(Player player) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+  return HistoryString();
+}
+
+std::string CheckersState::ObservationString(Player player) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+  return ToString();
+}
+
+void CheckersState::ObservationTensor(Player player,
+                                     absl::Span<float> values) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+
+  TensorView<3> view(values, {kCellStates, rows_, columns_},
+                               true);
+
+  // Observation Tensor Representation:
+  //   Plane 0: 1's where the current player's pieces are, 0's elsewhere.
+  //   Plane 1: 1's where the oppponent's pieces are, 0's elsewhere.
+  //   Plane 2: 1's where the current player's crowned pieces are, 0's elsewhere.
+  //   Plane 3: 1's where the oppponent's crowned pieces are, 0's elsewhere.
+  //   Plane 4: 1's where the empty cells are, 0's elsewhere.
+  for (int row = 0; row < rows_; row++) {
+    for (int column = 0; column < columns_; column++) {
+      int plane = ObservationPlane(BoardAt(row, column), player);
+      view[{plane, row, column}] = 1.0;
+    }
+  }
+}
+
+void CheckersState::UndoAction(Player player, Action action) {
+  std::vector<int> values =
+      UnrankActionMixedBase(action, {rows_, columns_, kNumDirections, kNumMoveType, kNumPieceType, kNumPieceType});
+
+  const int start_row = values[0];
+  const int start_column = values[1];
+  const int direction = values[2];
+  const int move_type = values[3];
+  const int captured_piece_type = values[4];
```

Comment:
Ah, I see, this is why the piece types are in there. Is it better to put more details in the history and undo based on that instead of including it in the action?

Reply:
Great suggestion. Have introduced a TurnHistoryInfo struct to do this. Now actions have reduced to 1/4th.